lmnet/lmnet/datasets/dataset_iterator.py

```diff
@@ -0,0 +1,331 @@
+from multiprocessing import Pool
+import time
+import sys
+import threading
+import numpy as np
+import os
+import queue
+from lmnet.datasets.base import SegmentationBase, ObjectDetectionBase
+
+
+_dataset = None
+
+
+def _prefetch_setup(dataset, seed, do_shuffle):
+    global _dataset
+    _dataset = dataset
+    if do_shuffle:
+        np.random.seed(os.getpid()+seed)
+        _dataset.seed = os.getpid() + seed
+        _dataset._shuffle()
+
+
+def _feed(i):
+    return _dataset.feed()
+
+
+def _apply_augmentations(dataset, image, label):
+    augmentor = dataset.augmentor
+    pre_processor = dataset.pre_processor
+
+    sample = {'image': image}
+
+    if issubclass(dataset.__class__, SegmentationBase):
+        sample['mask'] = label
+    elif issubclass(dataset.__class__, ObjectDetectionBase):
+        sample['gt_boxes'] = label
+    else:
+        sample['label'] = label
+
+    if callable(augmentor) and dataset.subset == 'train':
+        sample = augmentor(**sample)
+
+    if callable(pre_processor):
+        sample = pre_processor(**sample)
+
+    image = sample['image']
+
+    if issubclass(dataset.__class__, SegmentationBase):
+        label = sample['mask']
+    elif issubclass(dataset.__class__, ObjectDetectionBase):
+        label = sample['gt_boxes']
+    else:
+        label = sample['label']
+
+    # FIXME(tokunaga): dataset should not have thier own data format
+    if dataset.data_format == "NCHW":
+        image = np.transpose(image, [2, 0, 1])
+
+    return (image, label)
+
+
+def _process_one_data(i):
+    image, label = _dataset[i]
+    return _apply_augmentations(_dataset, image, label)
+
+
+def _concat_data(data_list):
+    images, labels = zip(*data_list)
+    images = np.array(images)
+    labels = np.array(labels)
+    return (images, labels)
+
+
+def _xorshift32(r):
+    r = r ^ (r << 13 & 0xFFFFFFFF)
+    r = r ^ (r >> 17 & 0xFFFFFFFF)
+    r = r ^ (r << 5 & 0xFFFFFFFF)
+    return r & 0xFFFFFFFF
+
+
+class _MultiProcessDatasetPrefetchThread(threading.Thread):
+    def __init__(self, dataset, result_queue, seed):
+        super().__init__()
+        # TODO(tokunaga): the number of processes should be configurable
+        self.seed = seed + 1  # seed must not be 0
+        self.support_getitem = hasattr(dataset, "__getitem__")
+        self.pool = Pool(processes=8, initializer=_prefetch_setup,
+                         initargs=(dataset, self.seed, not self.support_getitem))
+        self.result_queue = result_queue
+        self.batch_size = dataset.batch_size
+        self.dataset = dataset
+        self.data_ids = []
+        self.terminate = False
```

Comment:
I'm sorry to comment on this old PR, yet when reading this file, I have some questions.
- What's the purpose of this `self.terminate`, I think it keeps `False` always.

Reply:
@yd8534976 
Yes. `self.terminate` keep `False` unfortunately , It need to change to handle terminate.